### ISSUE TITLE
[Bug]`layernorm_quant` FP8 e4m3 correctness failure on A10/SM86 with small scale

### ISSUE THREAD

**Issue by cindyzxq** [flashinfer-ai/flashinfer#4160]
## Description

  `flashinfer.norm.layernorm_quant` added by #3962 appears to fail correctness on A10/SM86 in the public nightly pipeline. Observed in https://github.com/flashinfer-ai/flashinfer/actions/runs/30186569236 
  ```bash
  pytest -s tests/utils/test_norm.py

  Failing cases:

  tests/utils/test_norm.py::test_layernorm_quant[quant_dtype0-0.01-500-1]
  tests/utils/test_norm.py::test_layernorm_quant[quant_dtype0-0.01-1024-1]

  Parameters:

  batch_size = 1
  hidden_size = 500 / 1024
  quant_scale = 0.01
  quant_dtype = torch.float8_e4m3fn

  Failure:

  assert default_check([y_ref], [y], max_mismatch_pct=1.0, min_cos_sim=None)
```
  This means more than 1% of elements are not close under the default FP8 tolerance (rtol=0.1, atol=0.1).

  ## Why this looks SM86/A10 related

  layernorm_quant does not currently gate out SM86, so A10 appears to be an intended supported target.

  The suspicious path is the packed e4m3 conversion used by even hidden sizes. In include/flashinfer/trtllm/common/cudaTypeUtils.cuh, the float2 -> __nv_fp8x2_e4m3 conversion currently rounds through bf16 before converting to FP8:
```
  template <>
  __device__ inline __nv_fp8x2_e4m3 cuda_cast<__nv_fp8x2_e4m3, float2>(float2 val) {
    return __nv_fp8x2_e4m3(bf1622float2(float22bf162(val)));
  }
```
  The Python reference in tests/utils/test_norm.py only models the earlier bf16 rounding before applying scale:
```
  y = y.to(x.dtype).float() / scale
```
  With scale=0.01, small rounding differences are amplified by 100x and can cross FP8 e4m3 bucket boundaries. This likely explains why the failure shows up for float8_e4m3fn, small scale, and even hidden sizes (500, 1024) on A10/SM86.

  ## Suggested fix (By Codex)

  Change the packed e4m3 cast to convert directly from float2 to FP8, matching the implementation style already used in include/flashinfer/vec_dtypes.cuh:
```
  template <>
  __device__ inline __nv_fp8x2_e4m3 cuda_cast<__nv_fp8x2_e4m3, float2>(float2 val) {
    __nv_fp8x2_e4m3 out;
    out.__x = __nv_cvt_float2_to_fp8x2(val, __NV_SATFINITE, __NV_E4M3);
    return out;
  }
```
  It may also be worth doing the analogous check for e5m2 consistency, though the observed failures are e4m3 only.

  ## Test improvements (By Codex)

  Make test_layernorm_quant deterministic by setting a seed before generating inputs:
```
  torch.manual_seed(0)
```
  Add temporary mismatch diagnostics while validating the fix:
```
  af = y.float()
  rf = y_ref.float()
  close = torch.isclose(af, rf, rtol=0.1, atol=0.1)

  print("mismatch_pct", 100 * (~close).float().mean().item())
  print("max_abs_diff", (af - rf).abs().max().item())
  print("bad_ref_actual", torch.stack([rf[~close], af[~close]], dim=1)[:20])
```
  ## Nightly / jit-cache note
  The public nightly path installs flashinfer-jit-cache and runs with FLASHINFER_DISABLE_JIT=1 on A10/SM86. Current nightly/release jit-cache arch lists appear to include 8.0 and 8.9, but not 8.6.

  It may be worth adding 8.6 to the jit-cache arch list so A10 tests use an exact SM86 prebuilt artifact rather than a compatible Ampere artifact:

  7.5 8.0 8.6 8.9 9.0a ...

  Relevant files:

  - .github/workflows/nightly-release.yml
  - .github/workflows/release.yml

  ## Expected behavior

  Either:

  1. layernorm_quant should pass FP8 e4m3 correctness on SM86/A10, or
  2. the API/test should explicitly gate unsupported SMs if SM86 FP8 correctness is not intended to be supported.

**Comment by cindyzxq** [comment]
cc: @kahyunnam @elwhyjay @aleozlx

**Comment by elwhyjay** [comment]
@cindyzxq  Thanks for the detailed report. The diagnosis is correct I will send a PR that converts float2 to e4m3 directly and seeds the test.

**Comment by elwhyjay** [comment]
Two things I found while fixing it. The e5m2 packed cast already converts without the bf16 round trip, so only e4m3 needs the change, which matches the failures you saw. And the bf16 round trip is architecture independent, so this is not really SM86 specific. I reproduced it on an A6000 with a JIT-compiled SM86 build, 7 of 60 seeds over the 1% bound at hidden 500 and 1024, and 0 after the fix. The missing seed in the test is what made it intermittent rather than always failing.
On the jit-cache arch list, the gap is real but it is not what caused this: my repro used a native SM86 build and still failed, so adding 8.6 would not have caught it any earlier. Given the earlier wheel size limits I would rather keep it out of this PR.

**Comment by kahyunnam** [comment]
Closing as fixed by #4167
